eras/mary/impl/testlib/Test/Cardano/Ledger/Mary/Arbitrary.hs

```diff
@@ -89,9 +102,90 @@ instance Crypto c => Arbitrary (PolicyID c) where
+--   4. So, we need to ensure that at least the last of the offsets (AssetName offsets) do
+--      not exceed 65535.
+--   5. With `n` as the total number of assets, `p` the number of policy ids, the inequality to be satisfied is thus:
+--           8n -- Word64 asset quantities
+--        +  2n -- Word16 policy id offsets
+--        +  2n -- Word16 asset name offsets
+--        + 28p -- 28-byte policy ids
+--        + 32n -- 32-byte asset names (a maximum of 32 bytes)
+--        should be <= 65535, assuming the numer of policies to be maximal (i.e. equal to number of assets)
+--        65535 / 72 ~ 910.2 is the maximum number of triples to be safely generated.
+--        Or, in other words, 44n + 28p <= 65535
+--
+-- NOTE: There are some conditions due to which exceeding this number may not
+-- result in a guaranteed failure to compact without overflow, because, during compacting
+--   1. The asset names and policy ids are deduplicated
+--   2. Not all generated asset names are 32-bytes long
+-- But, exceeding this number does make the probability of causing overflow > 0.
+genMultiAsset :: forall c. Crypto c => Gen Integer -> Gen (MultiAsset c)
+genMultiAsset genAmount = do
+  MultiAsset ma <-
+    oneof
+      [ MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary genAmount)
+      , multiAssetFromListBounded <$> listOf1 (genMultiAssetTriple $ fromIntegral <$> genAmount)
+      ]
+  if 44 * sum (length <$> ma) + 28 * length ma > 65535
+    then scale (`div` 2) $ genMultiAsset genAmount
+    else pure $ MultiAsset ma
+
+-- | For tests, because `insertMultiAsset` called through `genMultiAsset` filters out zero values
+genMultiAssetZero :: Crypto c => Gen (MultiAsset c)
+genMultiAssetZero = MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary (pure 0))
+
+-- For negative tests, we need a definite generator that will produce just large-enough MultiAssets
+-- that will fail decoding, but not large-enough to consume too much resource.
+genMultiAssetToFail :: forall c. Crypto c => Gen (MultiAsset c)
+genMultiAssetToFail = do
+  let genAssetNameToFail = AssetName <$> genShortByteString 32
+  (numP, numA) <-
+    oneof
+      -- Ensure that #policies <= #assets by minimizing the policies and maximizing the assets.
+      [ do
+          numAssetNames <- chooseInt (1, 1489) -- n > 1489 will produce negative `minNumPolicyIds`
+          let minNumPolicyIds = (65535 - 44 * numAssetNames) `div` 28
+          numPolicyIds <- (minNumPolicyIds +) . getPositive <$> arbitrary
+          pure (min numPolicyIds numAssetNames, max numPolicyIds numAssetNames)
+      , do
+          numPolicyIds <- chooseInt (1, 2340) -- p > 2340 will produce negative `minNumAssetNames`
+          let minNumAssetNames = (65535 - 28 * numPolicyIds) `div` 44
+          numAssetNames <- (minNumAssetNames +) . getPositive <$> arbitrary
+          pure (min numPolicyIds numAssetNames, max numPolicyIds numAssetNames)
+      ]
```

Comment:
I think this is going to be the best strategy. I realized that there is no need to have two separate approaches for generating `numP` and `numA`, both of them rely on the same math, so one of them is redundant.
Also, I realized that when `minNumPolicyIds` is negative, it means that the number of asset names is big enough by itself to overflow the compaction, which means it sufficient to have just one policy that contains all of those asset names.
```suggestion
    do
          -- when numAssetNames > 1489 it is enough to have at least 1 PolicyId
          numAssetNames <- chooseInt (1, 1500)
          -- Make sure there is at least 1 PolicyId:
          let minNumPolicyIds = max 0 ((65535 - 44 * numAssetNames) `div` 28)
          numPolicyIds <- (minNumPolicyIds +) . getPositive <$> arbitrary
          -- Ensure we have at least as many asset names as there are PolicyIds
          pure (numPolicyIds, max numPolicyIds numAssetNames)
```